addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,766 @@
+        # on their system (timeout, for example)
+        if self.l10n_vn_edi_invoice_transaction_id:
+            invoice_lookup, error_message = self._l10n_vn_edi_lookup_invoice()
+            if 'result' in invoice_lookup:
+                invoice_data = invoice_lookup['result'][0]
+            # note: We do not catch errors on this endpoint for simplicity, as it should not be required.
+        else:
+            # We do not store the transaction id on the move right away so that we can avoid the above api call.
+            # When sending for the first time, we'll get the id from the file data, which we generated earlier in the flow.
+            self.l10n_vn_edi_invoice_transaction_id = invoice_json_data['generalInvoiceInfo']['transactionUuid']
+
+        # if the above request did not return data, we can assume that the invoice has failed to be created, or was never sent
+        if not invoice_data:
+            # Send the invoice to the system
+            request_response, error_message = self._l10n_vn_edi_send_request(
+                method='POST',
+                endpoint=f'InvoiceAPI/InvoiceWS/createInvoice/{self.company_id.vat}',
+                json_data=invoice_json_data,
+            )
+
+            if error_message:
+                return [error_message]
+
+            invoice_data = request_response['result']
+
+        self.write({
+            'l10n_vn_edi_reservation_code': invoice_data['reservationCode'],
+            'l10n_vn_edi_invoice_number': invoice_data['invoiceNo'],
+            'l10n_vn_edi_invoice_state': 'sent',
+        })
+
+    def _l10n_vn_edi_cancel_invoice(self, reason, agreement_document_name, agreement_document_date):
+        """ Send a request to cancel the invoice. """
+        self.ensure_one()
+
+        # Lock
+        self._l10n_vn_edi_lock_records()
+
+        # First, try to cancel the invoice inside Odoo
+        self._check_fiscalyear_lock_date()
+        self.line_ids._check_tax_lock_date()
+
+        self.button_draft()
+        self.button_cancel()
+
+        # If no error raised, we try to cancel it on the EDI.
+        _request_response, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='InvoiceAPI/InvoiceWS/cancelTransactionInvoice',
+            params={
+                'supplierTaxCode': self.company_id.vat,
+                'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+                'invoiceNo': self.l10n_vn_edi_invoice_number,
+                'strIssueDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+                'additionalReferenceDesc': agreement_document_name,
+                'additionalReferenceDate': self._l10n_vn_edi_format_date(agreement_document_date),
+                'reasonDelete': reason,
+            },
+        )
+        if error_message:
+            raise UserError(error_message)
```

Comment:
it's a bit weird to cancel the invoice, then cancel from the web-service, then rollback everything in case of error. What about calling the web-service first, then cancel the invoice with a try/except in case of locked period?

Reply:
I think it makes sense, we have more control over Odoo.
What if the invoice becomes cancelled on the edi platform, but then there is an error in Odoo because the invoice shouldn't be allowed to be cancelled?

Wouldn't you end up with an inconsistent state?
Here:

- If it fails while trying to cancel in Odoo, it will rollback (EDI not updated since it won't continue) :heavy_check_mark: 
- If it fails while trying to cancel on the platform, it will rollback (EDI not updated since failure) :heavy_check_mark: 
=> State stays consistent

If I change:
- If it fails while trying to cancel on the platform, it will rollback (EDI not updated since failure) :heavy_check_mark: 
- If it fails while trying to cancel in Odoo, it will rollback (EDI has already been updated) :question:

Reply:
@vin-odoo In any case, I think it's safer to commit explicitely after the EDI call. I would do something similar to Mexico:
- ask a cancellation
- if accepted, commit
- then:
```py
            try:
                self._check_fiscal_lock_dates()
                self.line_ids._check_tax_lock_date()

                self.button_draft()
                self.button_cancel()
            except UserError:
                pass
```
In case of locked period, let the invoice open and the accountant will add a credit note later (not send to the EDI). That's one of the rare case in which you don't want to send your credit note to the government